katran/lib/bpf/pckt_encap.h

```diff
@@ -218,19 +227,27 @@ static inline bool gue_encap_v4(struct xdp_md *xdp, struct ctl_value *cval,
   memcpy(new_eth->eth_source, old_eth->eth_dest, sizeof(new_eth->eth_source));
   new_eth->eth_proto = BE_ETH_P_IP;
 
+  if (is_ipv6) {
+    sport ^= ((pckt->flow.srcv6[3] >> 16) & 0xFFFF);
+    pkt_bytes += (sizeof(struct ipv6hdr) + sizeof(struct udphdr));
+  } else {
+    sport ^= ((pckt->flow.src >> 16) & 0xFFFF);
+    pkt_bytes += sizeof(struct udphdr);
+  }
+
   create_udp_hdr(
     udph,
     sport,
     GUE_DPORT,
-    pkt_bytes + sizeof(struct udphdr),
+    pkt_bytes,
     GUE_CSUM);
```

Comment:
Can you rebase? We've changed the computation of csum for gue.

Reply:
> You also need to update decap_* methods since the decap for ipv4 assume the inner packet to be v4 as well.

I have updated `(gue_)decap_(v4|v6)` methods in latest commit.



> Can you rebase? We've changed the computation of csum for gue.

updated.



In addition, there is a bug in `add_pseudo_ipv4_header` and `rem_pseudo_ipv4_header` in this commit (https://github.com/facebookincubator/katran/commit/b570f79a955e91a00bc80b208ce55892cc77e31c)

`iph->tot_len` contains the size of IPv4 header, but `udp length` field of pseudo header should be payload size (size  of (udp header + data) ).

It leads to udp checksum error when encap mode is IPv4-in-IPv6 (or IPv6-in-IPv4).

Reply:
Thanks for flagging. Yes, that looks like a bug indeed.